softlayer/resource_softlayer_network_vlan.go

```diff
@@ -0,0 +1,419 @@
+	}
+
+	if vlan.PrimaryRouter != nil {
+		d.Set("primary_router_hostname", *vlan.PrimaryRouter.Hostname)
+		if strings.HasPrefix(*vlan.PrimaryRouter.Hostname, "fcr") {
+			d.Set("type", "PUBLIC")
+		} else {
+			d.Set("type", "PRIVATE")
+		}
+		if vlan.PrimaryRouter.Datacenter != nil {
+			d.Set("datacenter", *vlan.PrimaryRouter.Datacenter.Name)
+		}
+	}
+
+	if vlan.BillingItem == nil {
+		d.Set("softlayer_managed", true)
+	} else {
+		d.Set("softlayer_managed", false)
+	}
+
+	// Subnets
+	subnets := make([]map[string]interface{}, 0)
+
+	for _, elem := range vlan.Subnets {
+		subnet := make(map[string]interface{})
+		subnet["subnet"] = *elem.NetworkIdentifier + "/" + strconv.Itoa(*elem.Cidr)
+		subnet["subnet_type"] = *elem.SubnetType
+		subnets = append(subnets, subnet)
+	}
+	d.Set("subnets", subnets)
+
+	if vlan.Subnets != nil && len(vlan.Subnets) > 0 {
+		d.Set("primary_subnet_size", 1<<(uint)(32-*vlan.Subnets[0].Cidr))
+	} else {
+		d.Set("primary_subnet_size", 0)
+	}
+
+	return nil
+}
+
+func resourceSoftLayerNetworkVlanUpdate(d *schema.ResourceData, meta interface{}) error {
+	sess := meta.(*session.Session)
+	service := services.GetNetworkVlanService(sess)
+
+	vlanId, err := strconv.Atoi(d.Id())
+	if err != nil {
+		return fmt.Errorf("Not a valid vlan ID, must be an integer: %s", err)
+	}
+
+	opts := datatypes.Network_Vlan{}
+
+	if d.HasChange("name") {
+		opts.Name = sl.String(d.Get("name").(string))
+	}
+
+	_, err = service.Id(vlanId).EditObject(&opts)
+
+	if err != nil {
+		return fmt.Errorf("Error updating Network Vlan: %s", err)
+	}
+	return resourceSoftLayerNetworkVlanRead(d, meta)
+}
+
+func resourceSoftLayerNetworkVlanDelete(d *schema.ResourceData, meta interface{}) error {
+	sess := meta.(*session.Session)
+	service := services.GetNetworkVlanService(sess)
+
+	vlanId, err := strconv.Atoi(d.Id())
+	if err != nil {
+		return fmt.Errorf("Not a valid vlan ID, must be an integer: %s", err)
+	}
+
+	// Check if the VLAN is existed.
+	_, err = service.Id(vlanId).Mask("id").GetObject()
+	if err != nil {
+		return fmt.Errorf("Error deleting Network Vlan: %s", err)
+	}
+
+	billingItem, err := service.Id(vlanId).GetBillingItem()
+	if err != nil {
+		return fmt.Errorf("Error deleting Network Vlan: %s", err)
+	}
+	if billingItem.Id == nil {
+		return nil
+	}
+
+	success, err := services.GetBillingItemService(sess).Id(*billingItem.Id).CancelService()
+	if err != nil {
+		return err
+	}
+
+	if !success {
+		return fmt.Errorf("SoftLayer reported an unsuccessful cancellation")
+	}
+
+	return nil
+}
+
+func resourceSoftLayerNetworkVlanExists(d *schema.ResourceData, meta interface{}) (bool, error) {
+	sess := meta.(*session.Session)
+	service := services.GetNetworkVlanService(sess)
+
+	vlanId, err := strconv.Atoi(d.Id())
+	if err != nil {
+		return false, fmt.Errorf("Not a valid vlan ID, must be an integer: %s", err)
+	}
+
+	_, err = service.Id(vlanId).Mask("id").GetObject()
+
+	if err != nil {
+		return false, err
+	}
+
+	return true, nil
```

Comment:
Does this logic work?  i.e., If Terraform receives a non-nil error from `...Exists`, does it even consider the first (`bool`) return value, or would the error disqualify the `exists` test?

Thinking it also doesn't handle non-404 errors (500, network, etc.)